Added e2e test for canary rollout

janeman98 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kubeflow/kfserving/blob/master/docs/DEVELOPER_GUIDE.md
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Add e2e test for canary rollout

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #538

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/kfserving/658)
<!-- Reviewable:end -->

janeman98 commented:
/assign @yuzisun

Dan, this test is built using the following as example:
- test_tensorflow.py 
- [sample for KFServing SDK](https://github.com/kubeflow/kfserving/blob/master/docs/samples/client/kfserving_sdk_sample.ipynb)

janeman98 commented:
/retest

ellis-bigelow commented:
Wow thank you for your PR. I don't believe we've met. Do you mind if I ask your background and interest in KFServing?

ellis-bigelow commented on test/e2e/predictor/test_canary.py:
```diff
@@ -0,0 +1,71 @@
+from ..common.utils import predict
+from ..common.utils import KFSERVING_TEST_NAMESPACE
+
+api_version = constants.KFSERVING_GROUP + '/' + constants.KFSERVING_VERSION
```
Is this API version not already in constants? If it's not, can you add it?

janeman98 replied:
Yes, already in https://github.com/kubeflow/kfserving/blob/master/python/kfserving/kfserving/constants/constants.py#L21

ellis-bigelow replied:
What i mean is, the constant file should already have this GROUP+VERSION concatenated string so you don't have to repeat this logic. Do you mind creating that as a new constant called constants.KFSERVING_API_VERSION so all tests can consume this?

janeman98 replied:
The constant file do not have  GROUP+VERSION concatenated string yet (currently they are two different separate strings) . Will add constants.KFSERVING_API_VERSION. Sorry for misunderstanding your previous comment :-)

ellis-bigelow commented on test/e2e/predictor/test_canary.py:
```diff
@@ -0,0 +1,71 @@
+from ..common.utils import KFSERVING_TEST_NAMESPACE
+
+api_version = constants.KFSERVING_GROUP + '/' + constants.KFSERVING_VERSION
+KFServing = KFServingClient(config_file="~/.kube/config")
```
As I understand it, the KFServingClient should default to this config file. Is this not the case?

janeman98 replied:
Yes, it is default to ~/.kube/config. Since SDK is running in a different cluster than KFServing, [user must set config_file to load kube-config file explicitly. ](https://github.com/kubeflow/kfserving/blob/master/python/kfserving/docs/KFServingClient.md)
`Note that for the case that the SDK is running in cluster and you want to operate KFServing in another remote cluster, user must set config_file to load kube-config file explicitly, e.g. KFServingClient(config_file="~/.kube/config").`

janeman98 replied:
Without doing this, presubmit test will fail.

ellis-bigelow commented on test/e2e/predictor/test_canary.py:
```diff
@@ -0,0 +1,71 @@
+                    limits={'cpu':'100m', 'memory':'256Mi'}))))
+
+
+    KFServing.rollout_canary(service_name, canary=canary_endpoint_spec, percent=10,
```
Does this call block? should we wait for isvc ready again?

janeman98 replied:
Correct. Will add.

ellis-bigelow commented on test/e2e/predictor/test_canary.py:
```diff
@@ -0,0 +1,71 @@
+       namespace=KFSERVING_TEST_NAMESPACE, watch=True, timeout_seconds=120)
+
+    # Promote Canary to Default
+    KFServing.promote(service_name, namespace=KFSERVING_TEST_NAMESPACE, watch=True, timeout_seconds=120)
```
likewise.

janeman98 replied:
Correct. Will add.

ellis-bigelow replied:
I think we should promote and then wait.  Your wait statements are all before, but I think they should be after. Currently this test doesn't validate that promote succeeds.

janeman98 replied:
Will add wait after promote.

janeman98 commented:
/retest

janeman98 commented:
/retest

janeman98 commented:
@ellis-bigelow 
Thanks for your review! I have addressed your review feedback and modified the code accordingly. Please review/approve. Feel free to let me know if further change is needed.

yuzisun commented on test/e2e/predictor/test_canary.py:
```diff
@@ -0,0 +1,73 @@
+       namespace=KFSERVING_TEST_NAMESPACE, watch=True, timeout_seconds=120)
+
+    # Promote Canary to Default
+    KFServing.wait_isvc_ready(service_name, namespace=KFSERVING_TEST_NAMESPACE)
```
I think this line should be after promote

yuzisun replied:
oh sorry actually need another wait after promote

janeman98 replied:
Oop, will add. Thanks Dan!

janeman98 commented:
@yuzisun @ellis-bigelow  

Thanks for the review! I have modified the source. Please review/approve. Feel free to let me know if further change is needed. Thanks again!

yuzisun commented:
/lgtm
/approve

yuzisun commented:
Thanks @janeman98 for adding the test!

janeman98 commented:
Thanks @yuzisun !
